### DISCUSSION TITLE
Global project-wide cs file for preprocessor instructions

### DISCUSSION THREAD

**Discussion by hwogit** [dotnet/csharplang#6534] [Language Ideas]
Instead of declaring namespaces project-wide in any cs file in C# 10, I would have found it better to optionally use a cs file that must have a specific name (e.g. Global.cs). It should be specified that this Global.cs may only contain preprocessor statements, namespace declarations and comments, but no C# code. This Global.cs, if present in the project, is automatically inserted by the compiler at the beginning of each cs file to be compiled.

2 reasons:

- You have a defined place where project-wide namespace declarations are declared.
- You have a central place where define switches can be set independent of the current platform (x86, x64, ARM, ARM64, AnyCPU, …) and independent of release or debug versions. This would be much more convenient than defining define statements in the project properties or in the csproj file. However, these could still be retained. In the Global.cs you could thus centrally describe define switches and set other switches depending on previously defined switches instead of repeating this in all necessary cs files. And you can comment preprocessor statements instead of just squeezing them into the csproj file.

Example Gobal.cs:
```

#if DEBUG
    // USE_TEST_DATA enables code that uses test data.
    #define USE_TEST_DATA
#else
    // PRESENTATION_MODE is commented here.
    #define PRESENTATION_MODE
#endif
```
The 2nd point is by far more important to me than the 1st point.

**Comment by svick** [comment]
Why is this significantly better than having it in the csproj file? For example, it could look like this:

```xml
<!-- USE_TEST_DATA enables code that uses test data. -->
<DefineConstants Condition="'$(Configuration)' == 'Debug'">$(DefineConstants);USE_TEST_DATA</DefineConstants>
<!-- PRESENTATION_MODE is commented here. -->
<DefineConstants Condition="'$(Configuration)' != 'Debug'">$(DefineConstants);PRESENTATION_MODE</DefineConstants>
```

Using MSBuild means the code is more verbose by quite a lot than it would be in C#, but I'm not sure that's an important factor here.

**Reply by hwogit** [reply]
The query on 'DEBUG' is only an example. There could also be if-queries on self-defined constants in the file 'Global.cs'. I think, if-queries in the csproj-file are not possible at all or only cumbersome.

I have never edited the csproj file in all my C# projects. I always use the Visual Studio dialog to do this. I just find it clearer to use the notation you use in a cs file rather than define constants in an XML formatted file. I edit almost exclusively code file types, xaml and cs.

You don't have to worry about which compiler configuration is currently enabled. One simply turns on or off a define switch in file 'Global.cs' to enable or disable code and thus can quickly add or disable code in any compiler configuration.

The 'DefineConstants' in the csproj file still have their meaning, for configuration-dependent switches that are set before the 'Global.cs' file is included.

**Reply by svick** [reply]
>  I think, if-queries in the csproj-file are not possible at all or only cumbersome.

They are possible. They might require some extra code, but I don't think it's that cumbersome.

> I have never edited the csproj file in all my C# projects. I always use the Visual Studio dialog to do this. 

That was definitely the norm in the .Net Framework days, since the csproj was an unreadable mess.

But I think that nowadays, it is acceptable that moderately advanced features require you to edit the csproj manually. Though maybe having to learn the intricacies of MSBuild is a step too far?

> You don't have to worry about which compiler configuration is currently enabled. One simply turns on or off a define switch in file 'Global.cs' to enable or disable code and thus can quickly add or disable code in any compiler configuration.

You can do the same thing in a csproj.

**Reply by hwogit** [reply]
I think, precompiler statements belong more to the code, not to the configuration.

It would be much more relaxed to write the precompiler statements in a cs file, without all the other things that belong in the csproj file.
